File: internal/ops/loop_status.go
package ops

import (
	"os"
	"strconv"
	"strings"
)

const (
	looprStatusStart = "---LOOPR_STATUS---"
	looprStatusEnd   = "---END_LOOPR_STATUS---"
)

type LooprStatus struct {
	Status           string
	ExitSignal       bool
	WorkType         string
	FilesModified    int
	ErrorCount       int
	Summary          string
	PermissionDenied bool
}

func ParseLooprStatusFromLog(path string) (LooprStatus, bool, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		return LooprStatus{}, false, err
	}
	status, ok := ParseLooprStatus(string(data))
	return status, ok, nil
}

func ParseLooprStatus(log string) (LooprStatus, bool) {
	idx := strings.LastIndex(log, looprStatusStart)
	if idx == -1 {
		return LooprStatus{}, false
	}
	segment := log[idx+len(looprStatusStart):]
	if end := strings.Index(segment, looprStatusEnd); end >= 0 {
		segment = segment[:end]
	}
	status := LooprStatus{
		Status: "UNKNOWN",
	}
	lines := strings.Split(segment, "\n")
	for _, line := range lines {
		line = strings.TrimSpace(line)
		if line == "" {
			continue
		}
		parts := strings.SplitN(line, ":", 2)
		if len(parts) != 2 {
			continue
		}
		key := strings.ToUpper(strings.TrimSpace(parts[0]))
		val := strings.TrimSpace(parts[1])
		switch key {
		case "STATUS":
			status.Status = strings.ToUpper(val)
		case "EXIT_SIGNAL":
			status.ExitSignal = parseBool(val)
		case "WORK_TYPE":
			status.WorkType = strings.ToLower(val)
		case "FILES_MODIFIED":
			status.FilesModified = parseInt(val)
		case "ERRORS":
			status.ErrorCount = parseInt(val)
		case "SUMMARY":
			status.Summary = val
		case "PERMISSION_DENIALS":
			status.PermissionDenied = parseBool(val)
		}
	}
	return status, true
}

func parseBool(value string) bool {
	switch strings.ToLower(strings.TrimSpace(value)) {
	case "true", "yes", "1", "y":
		return true
	default:
		return false
	}
}

func parseInt(value string) int {
	parsed, err := strconv.Atoi(strings.TrimSpace(value))
	if err != nil {
		return 0
	}
	return parsed
}

File: src/ops/loop_status.rs
use std::fs;
use std::path::Path;

use crate::{LooprError, LooprResult};

pub const LOOPR_STATUS_START: &str = "---LOOPR_STATUS---";
pub const LOOPR_STATUS_END: &str = "---END_LOOPR_STATUS---";

#[derive(Debug, Clone, Default)]
pub struct LooprStatus {
    pub status: String,
    pub exit_signal: bool,
    pub summary: String,
}

pub fn parse_loopr_status_from_log(path: &Path) -> LooprResult<(LooprStatus, bool)> {
    let data = fs::read_to_string(path)
        .map_err(|err| LooprError::new(format!("read {}: {}", path.display(), err)))?;
    Ok(parse_loopr_status(&data))
}

pub fn parse_loopr_status(log: &str) -> (LooprStatus, bool) {
    let idx = match log.rfind(LOOPR_STATUS_START) {
        Some(value) => value,
        None => return (LooprStatus::default(), false),
    };
    let mut segment = &log[idx + LOOPR_STATUS_START.len()..];
    if let Some(end) = segment.find(LOOPR_STATUS_END) {
        segment = &segment[..end];
    }

    let mut status = LooprStatus::default();
    for line in segment.lines() {
        let trimmed = line.trim();
        if trimmed.is_empty() {
            continue;
        }
        let mut parts = trimmed.splitn(2, ':');
        let key = match parts.next() {
            Some(value) => value.trim().to_uppercase(),
            None => continue,
        };
        let value = match parts.next() {
            Some(value) => value.trim(),
            None => continue,
        };
        match key.as_str() {
            "STATUS" => status.status = value.to_uppercase(),
            "EXIT_SIGNAL" => status.exit_signal = parse_bool(value),
            "SUMMARY" => status.summary = value.to_string(),
            _ => {}
        }
    }

    (status, true)
}

fn parse_bool(value: &str) -> bool {
    matches!(
        value.trim().to_lowercase().as_str(),
        "true" | "yes" | "1" | "y"
    )
}
